Improve external accounts system in custom profile fields

Niloth-p opened the pull request:
Fixes: #36461

The first 6 commits with prefix `settings_profile_fields` and `templates` are **refactor** commits.

### Functional changes
- Dropdown widgets support clicks on the sticky bottom option in calls to `render()`.
- The URL pattern in external accounts is now optional. Allowing the addition of text fields as external accounts.
- We use a filterable dropdown widget for choosing external account type.
- Existing external account types are not shown in the create form. (They previously used to be shown, but would raise an error upon attempting to create a new field)
- Added 19 new external accounts (all mentioned in the issue). And necessary icons.
- Default external account fields now support hints.

### Decisions

- This PR makes the "Custom" option stick to the bottom. To increase its visibility among our 20+ default accounts.
- The initial option used to be the "Custom" option. I have added "Select an account type" as the initial selected option of the dropdown. As the URL pattern, name and hint fields depend on the chosen account type, the initial form load now looks different. The URL pattern has been hidden, and the name and hint fields are disabled until the account type is chosen.
- Previously, hints were disabled for default account types. All of them had empty hint fields.
The last commit in this PR allows us to add hints for the default accounts. But I'm not sure if we'd like that or prefer to have hints disabled.

### URL Patterns for the new external accounts

Some of these accounts support some other URLs, but I've used the URL that they redirect to.
Examples:
- https://medium.com/@username -> https://username.medium.com/
- https://ca.linkedin.com/in/username -> https://linkedin.com/in/username
- https://www.snapchat.com/add/username -> https://www.snapchat.com/@username
- https://threads.net -> https://threads.com

YouTube supports both https://youtube.com/@handle and https://youtube.com/handle. I've kept the URL pattern as the former so that the link works even if the user includes the @ prefix in their handle value.

For Mastodon, we're accepting full URLs as the value.
For Discord, we're accepting the tags as text values. 

---

## Icons

- I've added attribution in THIRDPARTY file only for the TikTok icon obtained from https://fontawesomeicons.com/svg/icons/tiktok.
- I've cleaned up all the SVG files using SVGO, and a manual round.
- Screenshots of all the icons are included below in the "Comparing Icons" section.
- Some of the icons are smaller/larger than others. We might want to make them uniform, this has not been done in this PR yet.

#### Sources of the 19 icons

- 13 of the 19 new icons required were available in Font Awesome 4. 5 more are available in Font Awesome 6, but we don't use FA6. Unfortunately, most of these are now outdated, so I've added SVGs for these too.
- Downloaded official branding assets for 15 of the 19 new accounts, and 2 of the older accounts (GitHub, Twitter (X)).
- Got the **Bluesky** icon from https://commons.wikimedia.org/wiki/File:Bluesky_Logo.svg.
- The **TikTok** logo in their official assets only has a colored version in an AI file. I tried converting it to SVG using Inkscape. But, it renders as shown below in "TikTok icon comparison" section. So, I went with https://fontawesomeicons.com/svg/icons/tiktok. Attribution added.

#### FA4 icons that have not been replaced

- I have not added a **LinkedIn** icon, because the FA4 icon looks almost same as the current one, and LinkedIn's official assets only contain PNGs.
- I have not added a **Behance** icon, because I couldn't find any official assets. The [Commons one](https://commons.wikimedia.org/wiki/File:Behance_logo.svg) is also unsuitable. And the FA4 icon is the only version of their logo I could find, so I didn't replace it.

#### Attribution

I'm not sure when I should be adding attribution for images. Almost every single SVG has been edited when cleaning it.
I've removed `rect` / `path` elements from the SVGs of - Medium, Facebook, Twitch, Snapchat, Pinterest - to align with our monochrome usage, as those were using a middle element layer with a different fill color.
But, I've only added attribution for the TikTok icon since I was not able to get it from the official brand assets, nor use the FA4 icons.


<details>
<summary><h3>Comparing icons</h3></summary>

In the user profile view.
Icons in the left column are FA4 icons.
Icons in the right column are the added SVGs.

| Dark theme | Light theme |
|---|---|
| <img width="213" height="714" alt="Pasted image 20251217162927" src="https://github.com/user-attachments/assets/2d9e8af3-91ce-42d3-9e35-0a9799fe1047" /> | <img width="213" height="714" alt="Pasted image 20251217163037" src="https://github.com/user-attachments/assets/c037fd96-f0b8-4f21-9c9d-3dbb97087fa8" /> |

</details>

<details>
<summary><h4>TikTok icon comparison</h4></summary>

| TikTok's official AI asset | After convertion to SVG |
|---|---|
| ![tiktok](https://github.com/user-attachments/assets/2d40508e-7b6d-4e7e-aabd-a4da495aab72) | ![tiktok2](https://github.com/user-attachments/assets/2b7f2b63-ec91-4cef-b6ad-e3aec506fdb4) |


</details>

---

## Follow-ups to consider

- The `zulip-icon` class icons are not under the same alignment as the `fa` icons. Especially in the profile view, the `zulip-icon` icons are raised. The CSS needs to be tweaked somewhere. A simple vertical-alignment rule doesn't suffice, so I've left it for later.
- ~~Some of the icons are smaller/larger than others.~~ Fixed
- ~~While the last commit in this PR allows us to add hints for the default accounts, it still does not allow the user to edit them.~~
~~Would we also want to allow users to edit the hint of default account types? Irrespective of whether the hint is empty?~~ Thinking again, I've realised we would not want this. It defeats the purpose of having default accounts. The DB does not currently store hints for default external accounts.
- We seem to have a "Copy URL" button for link-type custom profile fields, do we also want it for external account fields? Could also be useful for something like the Discord tag and other text-only fields.
- Would we want to add the icons to the profile settings page too? We currently only show them in the user profile modal and user card popovers.

---

<details>
<summary><h3>Screenshots</h3></summary>

#### Creation form

**On choosing the field type**

<img width="530" height="536" alt="image" src="https://github.com/user-attachments/assets/33c1a620-1d61-472a-b275-66b99a65b9d8" />

**Opened dropdown**

<img width="530" height="536" alt="image" src="https://github.com/user-attachments/assets/62e3c344-d33e-4aed-8e74-c7fda9a91abf" />

Things to note:
- The first option "Custom"
- Default text "Select an account type"
- Existing external account profile fields are not available as options in the dropdown.

**Dropdown option selected: default external account**

<img width="660" height="666" alt="image" src="https://github.com/user-attachments/assets/74eaadc7-034a-4483-ba56-a24cc472fae5" />

**Dropdown option selected: custom external account**

<img width="625" height="734" alt="image" src="https://github.com/user-attachments/assets/b0a8f7fb-4b2a-4cd1-8d3b-8fddfe663abd" />


#### Edit form

<img width="660" height="590" alt="image" src="https://github.com/user-attachments/assets/82663a45-b4f3-426e-83b7-ced234396f67" />

#### Custom Profile Fields table

<img width="1310" height="1184" alt="image" src="https://github.com/user-attachments/assets/446e95dc-04ed-4f73-8756-7a4739c00404" />

#### Profile settings page

<img width="345" height="625" alt="img1" src="https://github.com/user-attachments/assets/74e555d3-1064-4ddc-a245-fc6cc228c4f4" />
<img width="355" height="861" alt="img2" src="https://github.com/user-attachments/assets/5b5f7668-e89a-4246-a8fa-e3d0b82785ef" />

#### User card popover

<img width="341" height="886" alt="img3" src="https://github.com/user-attachments/assets/6cb6a849-afdc-4deb-9e1c-4ef4d27b8899" />

Note: The Discord logo is not coloured because it's not a link.

#### User profile modal

<img width="226" height="776" alt="img4" src="https://github.com/user-attachments/assets/360d8331-fc08-4e5f-ba75-be7977631bb1" />
<img width="356" height="717" alt="img5" src="https://github.com/user-attachments/assets/784c8ce8-43e0-4e39-b3a9-74cbe207917c" />


</details>
<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on web/src/settings_profile_fields.ts:
```diff
@@ -157,11 +198,46 @@ function initialize_form_to_defaults(): void {
+        $("#custom_external_account_url_pattern").show();
+        $("#profile_field_name").val("").prop("disabled", false);
+        $("#profile_field_hint").val("").prop("disabled", false);
+        widget.render("Custom");
```
Ah, thank you for catching that!
I've now removed the commit extending `render` for `sticky_bottom_option` clicks. And made the "Custom" option the first option. I'll look into if we can add support for making `sticky_bottom_option` an actual option in `dropdown_widget.ts`, without breaking its current usage. Please let me know if you have any tips or ideas on how we can efficiently implement that.

Niloth-p replied:
The thing is we need to retain the current behavior of `sticky_bottom_option` as just text as well.
Currently, only saved snippets uses the sticky bottom option though.
I'll have to check if it would make sense to make that an option too.

Yes, that would be convenient, there's already much to discuss and review in this PR. I'll open that as a separate PR, when I implement it.

alya commented:
> Note that the GitLab and YouTube icons are much smaller compared to the rest.

How do they appear in actual usage? I don't think we want them to be tiny like this.

sahil839 commented on web/src/settings_components.ts:
```diff
@@ -442,12 +442,29 @@ export const external_account_field_schema = z.object({
-    };
+function read_external_account_field_data(
+    $profile_field_form: JQuery,
+    old_field_data: unknown,
```
Ok. Now I think `old_field_data` might be used in the function as we can directly return `subtype` from `old_field_data` for editing case since we do not allow changing subtype when editing the field. I think that might make the code more simple.

Niloth-p replied:
Good idea, thanks, done!

sahil839 commented on web/src/settings_profile_fields.ts:
```diff
@@ -822,17 +822,23 @@ function set_up_select_field(): void {
         const external_account = realm.realm_default_external_accounts[field_subtype];
         assert(external_account !== undefined);
         field_url_pattern = external_account.url_pattern;
+        // Use as text fields when there is no URL pattern
```
Ohk. I thought we only allowed custom external account fields without URL.

Niloth-p commented on web/src/settings_profile_fields.ts:
```diff
@@ -50,15 +50,31 @@ function setup_external_accounts_dropdown_widget(): void {
+                        JSON.parse(field.field_data),
+                    ),
+                )
+                .filter((fieldData) => fieldData.subtype && fieldData.subtype !== "custom")
```
Yes, we already check for it in the backend. We raise an error for that. Here's how it looks right now.

<img width="365" height="162" alt="img1" src="https://github.com/user-attachments/assets/aacffea9-9f61-4378-8d26-7dbf29df7ce9" />

So, instead of showing an error, we just remove existing accounts' options from the dropdown, in this commit.

A summarized version of this is also available in the PR description [here](https://github.com/zulip/zulip/pull/36811#:~:text=Existing%20external%20account%20types%20are%20not%20shown%20in%20the%20create%20form.%20(They%20previously%20used%20to%20be%20shown%2C%20but%20would%20raise%20an%20error%20upon%20attempting%20to%20create%20a%20new%20field)).
I've now added a line to the commit message.

Also, it's a separate commit since this is not about the select -> dropdown switch.

Niloth-p commented on web/src/settings_profile_fields.ts:
```diff
@@ -183,20 +249,9 @@ function update_form_for_field_type_selection(): void {
-                $("#profile_field_name").val(profile_field_name).prop("disabled", true);
-                $("#profile_field_hint").val("").prop("disabled", true);
-            }
+            external_accounts_dropdown_widget.render("default");
```
Yes, we don't need to pass an arg. It'll use the `default_id`. I had forgotten to update these `render()` calls in the rest of the file. Thanks for the reminder!

Niloth-p reviewed (commented):
Thanks for the review!

Updates:
- Rebased onto main, for conflict resolution.
- Reverted to using `old_field_data`, and updated the edit-field case to use it directly instead of fetching from `select` element.
- Updated the dropdown's `render()` calls.
- Added a few lines to a couple of commit messages for clarity, as requested.

sahil839 commented on web/src/settings_components.ts:
```diff
@@ -450,11 +450,12 @@ function read_external_account_field_data(
+        const widget = get_widget_for_dropdown_list_settings("external_accounts_type");
+        assert(widget !== null);
+        const value = widget.value();
+        assert(typeof value === "string");
```
Here I meant only the change to use widget value instead of text displayed in the option, since we use `widget.value()` for other cases as well to get the selected value in dropdown.

Niloth-p commented:
Updates:
- Updated the Twitter rename commit to include the migration, and renamed both the `subtype` and the `name`.
- Fixed the previously inconsistent sizes of icons.
- Squashed the commit that stashed the dropdown widget to get the external account value - into the commit adding the dropdown, as suggested.

timabbott commented:
This looks great! Merged, thanks @Niloth-p!

The alignment issue you noted seems worth debugging, maybe worth a #frontend thread if you're not sure what the bug is to get additional eyes on it.

> We seem to have a "Copy URL" button for link-type custom profile fields, do we also want it for external account fields? Could also be useful for something like the Discord tag and other text-only fields.

This seems worth prototyping and posting screenshots in #design for feedback.

> Would we want to add the icons to the profile settings page too? We currently only show them in the user profile modal and user card popovers.

Possibly! This seems worth prototyping and posting in #design for feedback as well.

One more follow-up to consider: Does this resolve https://github.com/zulip/zulip/issues/34392 through making the URL field optional? If not, we should probably look at that while this is fresh.
